File: ratings/src/main/java/com/cinelist/ms/ratings/handlers/RetrieveMessageErrorDecoder.java
package com.cinelist.ms.ratings.handlers;

import com.cinelist.ms.ratings.dtos.handler.CustomErrorResponse;
import com.cinelist.ms.ratings.handlers.exceptions.ResourceNotFoundException;
import com.fasterxml.jackson.databind.ObjectMapper;
import feign.Response;
import feign.codec.ErrorDecoder;
import org.apache.coyote.BadRequestException;

import java.io.IOException;
import java.io.InputStream;

public class RetrieveMessageErrorDecoder implements ErrorDecoder {
    private final ErrorDecoder errorDecoder = new Default();

    @Override
    public Exception decode(String methodKey, Response response) {
        CustomErrorResponse message = null;
        try (InputStream bodyIs = response.body().asInputStream()) {
            ObjectMapper mapper = new ObjectMapper();
            message = mapper.readValue(bodyIs, CustomErrorResponse.class);
        } catch (IOException ex) {
            return new Exception(ex.getMessage());
        }

        return switch (response.status()) {
            case 400 -> new BadRequestException(message.getError());
            case 404 -> new ResourceNotFoundException(message.getError());
            default -> errorDecoder.decode(methodKey, response);
        };
    }
}

File: accounts/src/main/kotlin/com/cinelits/ms/accounts/handlers/RetrieveMessageErrorDecoder.kt
package com.cinelits.ms.accounts.handlers

import com.cinelits.ms.accounts.dtos.handlers.CustomErrorResponse
import com.cinelits.ms.accounts.handlers.exceptions.ResourceNotFoundException
import com.fasterxml.jackson.databind.ObjectMapper
import feign.Response
import feign.codec.ErrorDecoder
import org.apache.coyote.BadRequestException
import java.io.IOException

class RetrieveMessageErrorDecoder : ErrorDecoder {
    private val errorDecoder: ErrorDecoder = ErrorDecoder.Default()

    override fun decode(methodKey: String?, response: Response): Exception {
        val message: CustomErrorResponse? = try {
            response.body().asInputStream().use { bodyIs ->
                val mapper = ObjectMapper()
                mapper.readValue(bodyIs, CustomErrorResponse::class.java)
            }
        } catch (ex: IOException) {
            return Exception(ex.message)
        }

        return when (response.status()) {
            400 -> BadRequestException(message?.message ?: "Bad Request")
            404 -> ResourceNotFoundException(message?.message ?: "Resource Not Found")
            else -> errorDecoder.decode(methodKey, response)
        }
    }
}